{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "02e78e94-348e-4897-ad4c-9197142669de",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cptac is already installed\n"
     ]
    }
   ],
   "source": [
    "# install the CPTAC helper package\n",
    "\n",
    "import importlib.util, sys, subprocess\n",
    "\n",
    "pkg = \"cptac\"\n",
    "if importlib.util.find_spec(pkg) is None:\n",
    "    print(\"Installing cptac...\")\n",
    "    subprocess.check_call([sys.executable, \"-m\", \"pip\", \"install\", \"cptac\"])\n",
    "else:\n",
    "    print(\"cptac is already installed\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "61e3952f-0fc7-4475-bc00-7e5b1bda908a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  Cancer            Source         Datatype\n",
      "0   ucec               bcm       proteomics\n",
      "1   ucec               bcm  transcriptomics\n",
      "2   None  brca_mapping.csv             None\n",
      "3   brca             broad  transcriptomics\n",
      "4  ccrcc             broad  transcriptomics\n",
      "5   coad             broad  transcriptomics\n",
      "6    gbm             broad  transcriptomics\n",
      "7  hnscc             broad  transcriptomics\n",
      "8   lscc             broad  transcriptomics\n",
      "9   luad             broad  transcriptomics\n"
     ]
    }
   ],
   "source": [
    "# load one CPTAC dataset\n",
    "\n",
    "import cptac\n",
    "# load one CPTAC datasets that got both proteomics and transcriptomics\n",
    "\n",
    "import cptac\n",
    "import pandas as pd\n",
    "\n",
    "df = cptac.list_datasets()\n",
    "print(df.head(10))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "b1beb26b-a271-43ae-8187-b2fee18fd6a9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Datatype                CNV  acetylproteomics  ancestry_prediction  cibersort  \\\n",
      "Cancer      Source                                                              \n",
      "all_cancers harmonized    0                 0                    1          0   \n",
      "            mssm          0                 0                    0          0   \n",
      "            washu         0                 0                    0          0   \n",
      "brca        bcm           1                 0                    0          0   \n",
      "            broad         0                 0                    0          0   \n",
      "            umich         0                 2                    0          0   \n",
      "            washu         1                 0                    0          1   \n",
      "ccrcc       bcm           1                 0                    0          0   \n",
      "            broad         0                 0                    0          0   \n",
      "            umich         0                 0                    0          0   \n",
      "\n",
      "Datatype                circular_RNA  clinical  hla_typing  mature_miRNA  \\\n",
      "Cancer      Source                                                         \n",
      "all_cancers harmonized             0         0           0             0   \n",
      "            mssm                   0         1           0             0   \n",
      "            washu                  0         0           1             0   \n",
      "brca        bcm                    0         0           0             0   \n",
      "            broad                  0         0           0             0   \n",
      "            umich                  0         0           0             0   \n",
      "            washu                  0         0           0             0   \n",
      "ccrcc       bcm                    1         0           0             0   \n",
      "            broad                  0         0           0             0   \n",
      "            umich                  0         0           0             0   \n",
      "\n",
      "Datatype                miRNA  phosphoproteomics  precursor_miRNA  proteomics  \\\n",
      "Cancer      Source                                                              \n",
      "all_cancers harmonized      0                  0                0           0   \n",
      "            mssm            0                  0                0           0   \n",
      "            washu           0                  0                0           0   \n",
      "brca        bcm             1                  1                0           1   \n",
      "            broad           0                  0                0           0   \n",
      "            umich           0                  1                0           1   \n",
      "            washu           0                  0                0           0   \n",
      "ccrcc       bcm             2                  2                0           2   \n",
      "            broad           0                  0                0           0   \n",
      "            umich           0                  1                0           1   \n",
      "\n",
      "Datatype                somatic_mutation  total_miRNA  transcriptomics  \\\n",
      "Cancer      Source                                                       \n",
      "all_cancers harmonized                 1            0                0   \n",
      "            mssm                       0            0                0   \n",
      "            washu                      0            0                0   \n",
      "brca        bcm                        0            0                1   \n",
      "            broad                      0            0                1   \n",
      "            umich                      0            0                0   \n",
      "            washu                      0            0                1   \n",
      "ccrcc       bcm                        0            0                1   \n",
      "            broad                      0            0                1   \n",
      "            umich                      0            0                0   \n",
      "\n",
      "Datatype                tumor_purity  xcell  \n",
      "Cancer      Source                           \n",
      "all_cancers harmonized             0      0  \n",
      "            mssm                   0      0  \n",
      "            washu                  1      0  \n",
      "brca        bcm                    0      0  \n",
      "            broad                  0      0  \n",
      "            umich                  0      0  \n",
      "            washu                  0      1  \n",
      "ccrcc       bcm                    0      0  \n",
      "            broad                  0      0  \n",
      "            umich                  0      0  \n",
      "Datatype       CNV  acetylproteomics  ancestry_prediction  cibersort  \\\n",
      "Cancer Source                                                          \n",
      "brca   bcm       1                 0                    0          0   \n",
      "ccrcc  bcm       1                 0                    0          0   \n",
      "coad   bcm       1                 0                    0          0   \n",
      "gbm    bcm       1                 0                    0          0   \n",
      "hnscc  bcm       0                 0                    0          0   \n",
      "lscc   bcm       1                 0                    0          0   \n",
      "luad   bcm       1                 0                    0          0   \n",
      "ov     bcm       1                 0                    0          0   \n",
      "pdac   bcm       1                 0                    0          0   \n",
      "ucec   bcm       1                 0                    0          0   \n",
      "\n",
      "Datatype       circular_RNA  clinical  hla_typing  mature_miRNA  miRNA  \\\n",
      "Cancer Source                                                            \n",
      "brca   bcm                0         0           0             0      1   \n",
      "ccrcc  bcm                1         0           0             0      2   \n",
      "coad   bcm                0         0           0             0      1   \n",
      "gbm    bcm                1         0           0             0      1   \n",
      "hnscc  bcm                1         0           0             0      2   \n",
      "lscc   bcm                1         0           0             0      2   \n",
      "luad   bcm                1         0           0             0      2   \n",
      "ov     bcm                0         0           0             0      0   \n",
      "pdac   bcm                1         0           0             0      2   \n",
      "ucec   bcm                1         0           0             0      1   \n",
      "\n",
      "Datatype       phosphoproteomics  precursor_miRNA  proteomics  \\\n",
      "Cancer Source                                                   \n",
      "brca   bcm                     1                0           1   \n",
      "ccrcc  bcm                     2                0           2   \n",
      "coad   bcm                     2                0           2   \n",
      "gbm    bcm                     1                0           1   \n",
      "hnscc  bcm                     2                0           2   \n",
      "lscc   bcm                     2                0           2   \n",
      "luad   bcm                     2                0           2   \n",
      "ov     bcm                     2                0           2   \n",
      "pdac   bcm                     2                0           2   \n",
      "ucec   bcm                     2                0           2   \n",
      "\n",
      "Datatype       somatic_mutation  total_miRNA  transcriptomics  tumor_purity  \\\n",
      "Cancer Source                                                                 \n",
      "brca   bcm                    0            0                1             0   \n",
      "ccrcc  bcm                    0            0                1             0   \n",
      "coad   bcm                    0            0                1             0   \n",
      "gbm    bcm                    0            0                1             0   \n",
      "hnscc  bcm                    0            0                1             0   \n",
      "lscc   bcm                    0            0                1             0   \n",
      "luad   bcm                    0            0                1             0   \n",
      "ov     bcm                    0            0                1             0   \n",
      "pdac   bcm                    0            0                1             0   \n",
      "ucec   bcm                    0            0                1             0   \n",
      "\n",
      "Datatype       xcell  \n",
      "Cancer Source         \n",
      "brca   bcm         0  \n",
      "ccrcc  bcm         0  \n",
      "coad   bcm         0  \n",
      "gbm    bcm         0  \n",
      "hnscc  bcm         0  \n",
      "lscc   bcm         0  \n",
      "luad   bcm         0  \n",
      "ov     bcm         0  \n",
      "pdac   bcm         0  \n",
      "ucec   bcm         0  \n"
     ]
    }
   ],
   "source": [
    "# Find the dataset with cancers, data source and the type of data\n",
    "\n",
    "df2 = df.dropna(subset=[\"Cancer\", \"Source\", \"Datatype\"]).copy()\n",
    "\n",
    "# Count datatypes per (Cancer, Source)\n",
    "counts = (\n",
    "    df2.groupby([\"Cancer\", \"Source\"])[\"Datatype\"]\n",
    "       .value_counts()\n",
    "       .unstack(fill_value=0)\n",
    ")\n",
    "\n",
    "print(counts.head(10))\n",
    "\n",
    "# Keep only Cancer/Source pairs that have both proteomics and transcriptomics\n",
    "both = counts[(counts.get(\"proteomics\", 0) > 0) & (counts.get(\"transcriptomics\", 0) > 0)]\n",
    "print(both)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "82829e49-5554-4db2-83d8-d9c08ab95bce",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Datatype columns (first 20):\n",
      "['CNV', 'acetylproteomics', 'ancestry_prediction', 'cibersort', 'circular_RNA', 'clinical', 'hla_typing', 'mature_miRNA', 'miRNA', 'phosphoproteomics', 'precursor_miRNA', 'proteomics', 'somatic_mutation', 'total_miRNA', 'transcriptomics', 'tumor_purity', 'xcell']\n",
      "\n",
      "Counts for ('ucec','bcm') (non-zero only):\n",
      "Datatype\n",
      "proteomics           2\n",
      "phosphoproteomics    2\n",
      "circular_RNA         1\n",
      "CNV                  1\n",
      "miRNA                1\n",
      "transcriptomics      1\n",
      "Name: (ucec, bcm), dtype: int64\n",
      "\n",
      "Specifically:\n",
      "proteomics: 2\n",
      "transcriptomics: 1\n"
     ]
    }
   ],
   "source": [
    "# confirm what datatypes exist for ucec / bcm\n",
    "\n",
    "import cptac\n",
    "df= cptac.list_datasets()\n",
    "df2= df.dropna(subset=[\"Cancer\", \"Source\", \"Datatype\"]).copy()\n",
    "\n",
    "counts= (df2.groupby([\"Cancer\", \"Source\"])[\"Datatype\"]\n",
    "         .value_counts()\n",
    "         .unstack(fill_value=0)\n",
    "        )\n",
    "  \n",
    "print(\"Datatype columns (first 20):\")\n",
    "print(list(counts.columns)[:20])\n",
    "\n",
    "print(\"\\nCounts for ('ucec','bcm') (non-zero only):\")\n",
    "row = counts.loc[(\"ucec\", \"bcm\")]\n",
    "print(row[row > 0].sort_values(ascending=False))\n",
    "\n",
    "print(\"\\nSpecifically:\")\n",
    "print(\"proteomics:\", int(row.get(\"proteomics\", 0)))\n",
    "print(\"transcriptomics:\", int(row.get(\"transcriptomics\", 0)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "id": "34e9c393-0760-4999-a3ab-bd368337e3a5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Ucec signature: (no_internet=False)\n",
      "\n",
      "Ucec docstring (first ~20 lines):\n",
      "\n",
      "    The Ucec class is a child class of the base Cancer class, intended for handling \n",
      "    Uterine Corpus Endometrial Carcinoma (UCEC) data from different sources.\n",
      "\n",
      "    Attributes:\n",
      "        _sources (dict): A dictionary of source objects that load and hold data \n",
      "                         pertaining to UCEC cancer.\n",
      "    \n",
      "Loaded dataset object: <class 'cptac.cancers.ucec.Ucec'>\n"
     ]
    }
   ],
   "source": [
    "# Load UCEC and inspect available tables\n",
    "\n",
    "import cptac\n",
    "import inspect\n",
    "\n",
    "print(\"Ucec signature:\", inspect.signature(cptac.Ucec))\n",
    "print(\"\\nUcec docstring (first ~20 lines):\")\n",
    "print(\"\\n\".join((cptac.Ucec.__doc__ or \"\").splitlines()[:20]))\n",
    "\n",
    "\n",
    "ds = cptac.Ucec()\n",
    "print(\"Loaded dataset object:\", type(ds))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "2f427cd4-93d4-4815-ac94-76c593f9a310",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Available get_* methods:\n",
      " - get_CNV\n",
      " - get_acetylproteomics\n",
      " - get_ancestry_prediction\n",
      " - get_cancer_type\n",
      " - get_cibersort\n",
      " - get_circular_RNA\n",
      " - get_clinical\n",
      " - get_data_list\n",
      " - get_dataframe\n",
      " - get_derived_molecular\n",
      " - get_docs\n",
      " - get_followup\n",
      " - get_genotype_all_vars\n",
      " - get_hla_typing\n",
      " - get_medical_history\n",
      " - get_miRNA\n",
      " - get_phosphoproteomics\n",
      " - get_proteomics\n",
      " - get_somatic_mutation\n",
      " - get_somatic_mutation_binary\n",
      " - get_targeted_phosphoproteomics\n",
      " - get_targeted_proteomics\n",
      " - get_transcriptomics\n",
      " - get_tumor_purity\n",
      " - get_xcell\n"
     ]
    }
   ],
   "source": [
    "methods = [m for m in dir(ds) if m.startswith(\"get_\")]\n",
    "print(\"Available get_* methods:\")\n",
    "for m in sorted(methods):\n",
    "    print(\" -\", m)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c3f30b33-e32b-44a3-a3df-883dc6a0bf2c",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
