{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "02e78e94-348e-4897-ad4c-9197142669de",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cptac is already installed\n"
     ]
    }
   ],
   "source": [
    "# install the CPTAC helper package\n",
    "\n",
    "import importlib.util, sys, subprocess\n",
    "\n",
    "pkg = \"cptac\"\n",
    "if importlib.util.find_spec(pkg) is None:\n",
    "    print(\"Installing cptac...\")\n",
    "    subprocess.check_call([sys.executable, \"-m\", \"pip\", \"install\", \"cptac\"])\n",
    "    subprocess.check_call([sys.executable, \"-m\", \"pip\", \"install\", \"-U\", \"cptac\"])\n",
    "\n",
    "else:\n",
    "    print(\"cptac is already installed\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "61e3952f-0fc7-4475-bc00-7e5b1bda908a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  Cancer            Source         Datatype\n",
      "0   ucec               bcm       proteomics\n",
      "1   ucec               bcm  transcriptomics\n",
      "2   None  brca_mapping.csv             None\n",
      "3   brca             broad  transcriptomics\n",
      "4  ccrcc             broad  transcriptomics\n",
      "5   coad             broad  transcriptomics\n",
      "6    gbm             broad  transcriptomics\n",
      "7  hnscc             broad  transcriptomics\n",
      "8   lscc             broad  transcriptomics\n",
      "9   luad             broad  transcriptomics\n"
     ]
    }
   ],
   "source": [
    "# load one CPTAC dataset\n",
    "\n",
    "import cptac\n",
    "# load one CPTAC datasets that got both proteomics and transcriptomics\n",
    "\n",
    "import cptac\n",
    "import pandas as pd\n",
    "\n",
    "df = cptac.list_datasets()\n",
    "print(df.head(10))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "b1beb26b-a271-43ae-8187-b2fee18fd6a9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Datatype                CNV  acetylproteomics  ancestry_prediction  cibersort  \\\n",
      "Cancer      Source                                                              \n",
      "all_cancers harmonized    0                 0                    1          0   \n",
      "            mssm          0                 0                    0          0   \n",
      "            washu         0                 0                    0          0   \n",
      "brca        bcm           1                 0                    0          0   \n",
      "            broad         0                 0                    0          0   \n",
      "            umich         0                 2                    0          0   \n",
      "            washu         1                 0                    0          1   \n",
      "ccrcc       bcm           1                 0                    0          0   \n",
      "            broad         0                 0                    0          0   \n",
      "            umich         0                 0                    0          0   \n",
      "\n",
      "Datatype                circular_RNA  clinical  hla_typing  mature_miRNA  \\\n",
      "Cancer      Source                                                         \n",
      "all_cancers harmonized             0         0           0             0   \n",
      "            mssm                   0         1           0             0   \n",
      "            washu                  0         0           1             0   \n",
      "brca        bcm                    0         0           0             0   \n",
      "            broad                  0         0           0             0   \n",
      "            umich                  0         0           0             0   \n",
      "            washu                  0         0           0             0   \n",
      "ccrcc       bcm                    1         0           0             0   \n",
      "            broad                  0         0           0             0   \n",
      "            umich                  0         0           0             0   \n",
      "\n",
      "Datatype                miRNA  phosphoproteomics  precursor_miRNA  proteomics  \\\n",
      "Cancer      Source                                                              \n",
      "all_cancers harmonized      0                  0                0           0   \n",
      "            mssm            0                  0                0           0   \n",
      "            washu           0                  0                0           0   \n",
      "brca        bcm             1                  1                0           1   \n",
      "            broad           0                  0                0           0   \n",
      "            umich           0                  1                0           1   \n",
      "            washu           0                  0                0           0   \n",
      "ccrcc       bcm             2                  2                0           2   \n",
      "            broad           0                  0                0           0   \n",
      "            umich           0                  1                0           1   \n",
      "\n",
      "Datatype                somatic_mutation  total_miRNA  transcriptomics  \\\n",
      "Cancer      Source                                                       \n",
      "all_cancers harmonized                 1            0                0   \n",
      "            mssm                       0            0                0   \n",
      "            washu                      0            0                0   \n",
      "brca        bcm                        0            0                1   \n",
      "            broad                      0            0                1   \n",
      "            umich                      0            0                0   \n",
      "            washu                      0            0                1   \n",
      "ccrcc       bcm                        0            0                1   \n",
      "            broad                      0            0                1   \n",
      "            umich                      0            0                0   \n",
      "\n",
      "Datatype                tumor_purity  xcell  \n",
      "Cancer      Source                           \n",
      "all_cancers harmonized             0      0  \n",
      "            mssm                   0      0  \n",
      "            washu                  1      0  \n",
      "brca        bcm                    0      0  \n",
      "            broad                  0      0  \n",
      "            umich                  0      0  \n",
      "            washu                  0      1  \n",
      "ccrcc       bcm                    0      0  \n",
      "            broad                  0      0  \n",
      "            umich                  0      0  \n",
      "Datatype       CNV  acetylproteomics  ancestry_prediction  cibersort  \\\n",
      "Cancer Source                                                          \n",
      "brca   bcm       1                 0                    0          0   \n",
      "ccrcc  bcm       1                 0                    0          0   \n",
      "coad   bcm       1                 0                    0          0   \n",
      "gbm    bcm       1                 0                    0          0   \n",
      "hnscc  bcm       0                 0                    0          0   \n",
      "lscc   bcm       1                 0                    0          0   \n",
      "luad   bcm       1                 0                    0          0   \n",
      "ov     bcm       1                 0                    0          0   \n",
      "pdac   bcm       1                 0                    0          0   \n",
      "ucec   bcm       1                 0                    0          0   \n",
      "\n",
      "Datatype       circular_RNA  clinical  hla_typing  mature_miRNA  miRNA  \\\n",
      "Cancer Source                                                            \n",
      "brca   bcm                0         0           0             0      1   \n",
      "ccrcc  bcm                1         0           0             0      2   \n",
      "coad   bcm                0         0           0             0      1   \n",
      "gbm    bcm                1         0           0             0      1   \n",
      "hnscc  bcm                1         0           0             0      2   \n",
      "lscc   bcm                1         0           0             0      2   \n",
      "luad   bcm                1         0           0             0      2   \n",
      "ov     bcm                0         0           0             0      0   \n",
      "pdac   bcm                1         0           0             0      2   \n",
      "ucec   bcm                1         0           0             0      1   \n",
      "\n",
      "Datatype       phosphoproteomics  precursor_miRNA  proteomics  \\\n",
      "Cancer Source                                                   \n",
      "brca   bcm                     1                0           1   \n",
      "ccrcc  bcm                     2                0           2   \n",
      "coad   bcm                     2                0           2   \n",
      "gbm    bcm                     1                0           1   \n",
      "hnscc  bcm                     2                0           2   \n",
      "lscc   bcm                     2                0           2   \n",
      "luad   bcm                     2                0           2   \n",
      "ov     bcm                     2                0           2   \n",
      "pdac   bcm                     2                0           2   \n",
      "ucec   bcm                     2                0           2   \n",
      "\n",
      "Datatype       somatic_mutation  total_miRNA  transcriptomics  tumor_purity  \\\n",
      "Cancer Source                                                                 \n",
      "brca   bcm                    0            0                1             0   \n",
      "ccrcc  bcm                    0            0                1             0   \n",
      "coad   bcm                    0            0                1             0   \n",
      "gbm    bcm                    0            0                1             0   \n",
      "hnscc  bcm                    0            0                1             0   \n",
      "lscc   bcm                    0            0                1             0   \n",
      "luad   bcm                    0            0                1             0   \n",
      "ov     bcm                    0            0                1             0   \n",
      "pdac   bcm                    0            0                1             0   \n",
      "ucec   bcm                    0            0                1             0   \n",
      "\n",
      "Datatype       xcell  \n",
      "Cancer Source         \n",
      "brca   bcm         0  \n",
      "ccrcc  bcm         0  \n",
      "coad   bcm         0  \n",
      "gbm    bcm         0  \n",
      "hnscc  bcm         0  \n",
      "lscc   bcm         0  \n",
      "luad   bcm         0  \n",
      "ov     bcm         0  \n",
      "pdac   bcm         0  \n",
      "ucec   bcm         0  \n"
     ]
    }
   ],
   "source": [
    "# Find the dataset with cancers, data source and the type of data\n",
    "\n",
    "df2 = df.dropna(subset=[\"Cancer\", \"Source\", \"Datatype\"]).copy()\n",
    "\n",
    "# Count datatypes per (Cancer, Source)\n",
    "counts = (\n",
    "    df2.groupby([\"Cancer\", \"Source\"])[\"Datatype\"]\n",
    "       .value_counts()\n",
    "       .unstack(fill_value=0)\n",
    ")\n",
    "\n",
    "print(counts.head(10))\n",
    "\n",
    "# Keep only Cancer/Source pairs that have both proteomics and transcriptomics\n",
    "both = counts[(counts.get(\"proteomics\", 0) > 0) & (counts.get(\"transcriptomics\", 0) > 0)]\n",
    "print(both)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "82829e49-5554-4db2-83d8-d9c08ab95bce",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Datatype columns (first 20):\n",
      "['CNV', 'acetylproteomics', 'ancestry_prediction', 'cibersort', 'circular_RNA', 'clinical', 'hla_typing', 'mature_miRNA', 'miRNA', 'phosphoproteomics', 'precursor_miRNA', 'proteomics', 'somatic_mutation', 'total_miRNA', 'transcriptomics', 'tumor_purity', 'xcell']\n",
      "\n",
      "Counts for ('ucec','bcm') (non-zero only):\n",
      "Datatype\n",
      "proteomics           2\n",
      "phosphoproteomics    2\n",
      "circular_RNA         1\n",
      "CNV                  1\n",
      "miRNA                1\n",
      "transcriptomics      1\n",
      "Name: (ucec, bcm), dtype: int64\n",
      "\n",
      "Specifically:\n",
      "proteomics: 2\n",
      "transcriptomics: 1\n"
     ]
    }
   ],
   "source": [
    "# confirm what datatypes exist for ucec / bcm\n",
    "\n",
    "import cptac\n",
    "df= cptac.list_datasets()\n",
    "df2= df.dropna(subset=[\"Cancer\", \"Source\", \"Datatype\"]).copy()\n",
    "\n",
    "counts= (df2.groupby([\"Cancer\", \"Source\"])[\"Datatype\"]\n",
    "         .value_counts()\n",
    "         .unstack(fill_value=0)\n",
    "        )\n",
    "  \n",
    "print(\"Datatype columns (first 20):\")\n",
    "print(list(counts.columns)[:20])\n",
    "\n",
    "print(\"\\nCounts for ('ucec','bcm') (non-zero only):\")\n",
    "row = counts.loc[(\"ucec\", \"bcm\")]\n",
    "print(row[row > 0].sort_values(ascending=False))\n",
    "\n",
    "print(\"\\nSpecifically:\")\n",
    "print(\"proteomics:\", int(row.get(\"proteomics\", 0)))\n",
    "print(\"transcriptomics:\", int(row.get(\"transcriptomics\", 0)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "34e9c393-0760-4999-a3ab-bd368337e3a5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Ucec signature: (no_internet=False)\n",
      "\n",
      "Ucec docstring (first ~20 lines):\n",
      "\n",
      "    The Ucec class is a child class of the base Cancer class, intended for handling \n",
      "    Uterine Corpus Endometrial Carcinoma (UCEC) data from different sources.\n",
      "\n",
      "    Attributes:\n",
      "        _sources (dict): A dictionary of source objects that load and hold data \n",
      "                         pertaining to UCEC cancer.\n",
      "    \n",
      "Loaded dataset object: <class 'cptac.cancers.ucec.Ucec'>\n"
     ]
    }
   ],
   "source": [
    "# Load UCEC and inspect available tables\n",
    "\n",
    "import cptac\n",
    "import inspect\n",
    "\n",
    "print(\"Ucec signature:\", inspect.signature(cptac.Ucec))\n",
    "print(\"\\nUcec docstring (first ~20 lines):\")\n",
    "print(\"\\n\".join((cptac.Ucec.__doc__ or \"\").splitlines()[:20]))\n",
    "\n",
    "\n",
    "ds = cptac.Ucec()\n",
    "print(\"Loaded dataset object:\", type(ds))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "2f427cd4-93d4-4815-ac94-76c593f9a310",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Available get_* methods:\n",
      " - get_CNV\n",
      " - get_acetylproteomics\n",
      " - get_ancestry_prediction\n",
      " - get_cancer_type\n",
      " - get_cibersort\n",
      " - get_circular_RNA\n",
      " - get_clinical\n",
      " - get_data_list\n",
      " - get_dataframe\n",
      " - get_derived_molecular\n",
      " - get_docs\n",
      " - get_followup\n",
      " - get_genotype_all_vars\n",
      " - get_hla_typing\n",
      " - get_medical_history\n",
      " - get_miRNA\n",
      " - get_phosphoproteomics\n",
      " - get_proteomics\n",
      " - get_somatic_mutation\n",
      " - get_somatic_mutation_binary\n",
      " - get_targeted_phosphoproteomics\n",
      " - get_targeted_proteomics\n",
      " - get_transcriptomics\n",
      " - get_tumor_purity\n",
      " - get_xcell\n"
     ]
    }
   ],
   "source": [
    "methods = [m for m in dir(ds) if m.startswith(\"get_\")]\n",
    "print(\"Available get_* methods:\")\n",
    "for m in sorted(methods):\n",
    "    print(\" -\", m)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "c3f30b33-e32b-44a3-a3df-883dc6a0bf2c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "UCEC clinical sources: []\n",
      "UCEC proteomics sources: ['bcm', 'umich']\n",
      "UCEC transcriptomics sources: ['bcm', 'broad', 'washu']\n"
     ]
    }
   ],
   "source": [
    "import cptac\n",
    "\n",
    "meta = cptac.list_datasets()\n",
    "\n",
    "def sources_for(cancer, dtype):\n",
    "    out = meta[(meta[\"Cancer\"] == cancer) & (meta[\"Datatype\"] == dtype)][\"Source\"]\n",
    "    return sorted([s for s in out.dropna().unique()])\n",
    "\n",
    "print(\"UCEC clinical sources:\", sources_for(\"ucec\", \"clinical\"))\n",
    "print(\"UCEC proteomics sources:\", sources_for(\"ucec\", \"proteomics\"))\n",
    "print(\"UCEC transcriptomics sources:\", sources_for(\"ucec\", \"transcriptomics\"))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "360e8df1-8c5a-4aa4-84b7-61540b702920",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/mnt/vol1/protein_cancer/envs/python_ai_env/bin/python\n",
      "Retry session set up\n"
     ]
    }
   ],
   "source": [
    "\n",
    "import sys\n",
    "print(sys.executable)\n",
    "\n",
    "# Adding automatic retry for Zenodo downloads\n",
    "import requests\n",
    "from requests.adapters import HTTPAdapter\n",
    "from urllib3.util.retry import Retry\n",
    "import cptac.tools.download_tools as download_tools\n",
    "\n",
    "retry = Retry(\n",
    "    total=12,\n",
    "    backoff_factor=2,\n",
    "    status_forcelist=[429, 500, 502, 503, 504],\n",
    "    allowed_methods=[\"GET\"],\n",
    "    respect_retry_after_header=True,\n",
    ")\n",
    "\n",
    "session = requests.Session()\n",
    "session.mount(\"https://\", HTTPAdapter(max_retries=retry))\n",
    "\n",
    "# This makes cptac's download code to use retrying session\n",
    "download_tools.requests = session\n",
    "\n",
    "print(\"Retry session set up\")\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "d422206b-2d4a-4f54-a0f4-2ffd21f5f5f5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'bcm': ['CNV',\n",
       "  'circular_RNA',\n",
       "  'miRNA',\n",
       "  'phosphoproteomics',\n",
       "  'proteomics',\n",
       "  'transcriptomics'],\n",
       " 'broad': ['transcriptomics'],\n",
       " 'mssm': ['clinical', 'follow-up', 'medical_history'],\n",
       " 'umich': ['acetylproteomics', 'phosphoproteomics', 'proteomics'],\n",
       " 'washu': ['CNV',\n",
       "  'cibersort',\n",
       "  'hla_typing',\n",
       "  'miRNA',\n",
       "  'somatic_mutation',\n",
       "  'transcriptomics',\n",
       "  'tumor_purity',\n",
       "  'xcell'],\n",
       " 'harmonized': ['ancestry_prediction', 'somatic_mutation']}"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load UCEC and find which sources actually have clinical data\n",
    "\n",
    "import cptac\n",
    "\n",
    "ds = cptac.Ucec()\n",
    "\n",
    "# list what each source can load\n",
    "available = {src: sorted(obj.load_functions.keys()) for src, obj in ds._sources.items()}\n",
    "available\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "c8b3b36c-f4ad-45e4-a7b6-3fba7f89c41a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Clinical source for UCEC: mssm\n"
     ]
    }
   ],
   "source": [
    "# pick the first source that supports clinical\n",
    "\n",
    "def first_source_with(ds, dtype):\n",
    "    for src, obj in ds._sources.items():\n",
    "        if dtype in obj.load_functions:\n",
    "            return src\n",
    "    return None\n",
    "\n",
    "clinical_source = first_source_with(ds, \"clinical\")\n",
    "print(\"Clinical source for UCEC:\", clinical_source)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "57651fc4-b682-40d7-b770-7495302bba67",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Download with a simple backoff loop\n",
    "\n",
    "import time\n",
    "\n",
    "def call_with_backoff(func, *args, **kwargs):\n",
    "    for attempt in range(1, 11):\n",
    "        try:\n",
    "            return func(*args, **kwargs)\n",
    "        except Exception as e:\n",
    "            if \"429\" in str(e):\n",
    "                wait = 30 * attempt\n",
    "                print(f\"Hit Zenodo rate limit (429). Waiting {wait}s then retrying...\")\n",
    "                time.sleep(wait)\n",
    "            else:\n",
    "                raise\n",
    "    raise RuntimeError(\"Still rate-limited after retries.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "ee3dbc42-48c1-48e7-81aa-3e498cf9f699",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Downloading clinical_Pan-cancer.May2022.tsv.gz: 100%|███| 243k/243k [00:01<00:00, 138kB/s]\n",
      "Downloading UCEC_proteomics_gene_abundance_log2_reference_intensity_normalized_Tumor.txt.g\n",
      "Downloading UCEC_proteomics_gene_abundance_log2_reference_intensity_normalized_Normal.txt.\n",
      "Downloading gencode.v34.basic.annotation-mapping.txt.gz: 100%|█| 1.75M/1.75M [00:14<00:00,\n",
      "Downloading UCEC-gene_rsem_removed_circRNA_tumor_normal_UQ_log2(x+1)_BCM.txt.gz: 100%|█| 8\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "clinical: (103, 124)\n",
      "proteomics: (113, 11949)\n",
      "transcriptomics: (119, 59286)\n"
     ]
    }
   ],
   "source": [
    "# Get the data and save it \n",
    "\n",
    "clinical = None\n",
    "if clinical_source is not None:\n",
    "    clinical = call_with_backoff(ds.get_clinical, source=clinical_source)\n",
    "\n",
    "proteomics = call_with_backoff(ds.get_proteomics, source=\"bcm\")\n",
    "transcriptomics = call_with_backoff(ds.get_transcriptomics, source=\"bcm\")\n",
    "\n",
    "print(\"clinical:\", None if clinical is None else clinical.shape)\n",
    "print(\"proteomics:\", proteomics.shape)\n",
    "print(\"transcriptomics:\", transcriptomics.shape)\n",
    "\n",
    "\n",
    "from pathlib import Path\n",
    "\n",
    "PROJECT = Path(\"/mnt/vol1/Multi-Omics-Cancer-Subtype-Discovery\")\n",
    "RAW = PROJECT / \"data\" / \"raw\"\n",
    "RAW.mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "if clinical is not None:\n",
    "    clinical.to_csv(RAW / \"clinical.csv\")\n",
    "\n",
    "proteomics.to_csv(RAW / \"proteomics.csv\")\n",
    "transcriptomics.to_csv(RAW / \"transcriptomics.csv\")\n",
    "\n",
    "print(\"Saved files:\")\n",
    "for p in sorted(RAW.glob(\"*.csv\")):\n",
    "    print(\" -\", p.name)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ce8714c8-675f-458c-b903-b7e0ee52088f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
