---
title: "batch detection"
date: "`r Sys.Date()`"
output: 
  html_document:
    theme: cosmo
    toc: true
    toc_float: true
    number_sections: true
    code_folding: hide
---

# Introduction

RNA-seq tables often include thousands of genes that barely change across samples (near-constant). Those genes:

add noise,

slow computation,

can dominate distance calculations in subtle ways.

So we drop genes with almost zero variance across the 95 tumour samples.

# Setup and Dependencies

```{r packages, message=FALSE, warning=FALSE, echo=TRUE}
suppressPackageStartupMessages({
library("purrr")
library("metafor")
library("dplyr")
library("readxl")
library("ggplot2")
library("tidyverse")
library("readr")
library("DESeq2")
library("dplyr")
library("stringr")
library("biomaRt")
  
})
```


# 

```{r, message=FALSE, warning=FALSE, echo=TRUE}

processed_path <- "/mnt/vol1/Multi-Omics-Cancer-Subtype-Discovery/data/processed"

prot <- readRDS(file.path(processed_path, "proteomics_qc_imputed.rds"))
rna  <- readRDS(file.path(processed_path, "rna_filtered_imputed.rds"))
clin <- readRDS(file.path(processed_path, "clinical_a.rds"))

# align samples
common <- intersect(intersect(rownames(prot), rownames(rna)), rownames(clin))
prot <- prot[common, , drop = FALSE]
rna  <- rna[common,  , drop = FALSE]
clin <- clin[common, , drop = FALSE]


```

# PCA: proteomics only, RNA only, combined


```{r gtex, message=FALSE, warning=FALSE, echo=TRUE}

# Remove genes with zero variance (constant columns)
gene_sd <- apply(rna, 2, sd, na.rm = TRUE)
rna2 <- rna[, gene_sd > 0, drop = FALSE]

print(paste("Genes before:", ncol(rna)))
print(paste("Genes after removing sd=0:", ncol(rna2)))

# Proteomics PCA
pca_prot <- prcomp(scale(as.matrix(prot)), center = FALSE, scale. = FALSE)
pcs_prot <- as.data.frame(pca_prot$x[, 1:2])
colnames(pcs_prot) <- c("PC1", "PC2")

# RNA PCA
pca_rna <- prcomp(scale(as.matrix(rna2)), center = FALSE, scale. = FALSE)
pcs_rna <- as.data.frame(pca_rna$x[, 1:2])
colnames(pcs_rna) <- c("PC1", "PC2")


# columns containing batch words
batch_words <- c("batch","plate","run","center","site","institution","lab","source","aliquot","tmt","channel")

batch_cols <- names(clin)[
  sapply(names(clin), function(nm) any(sapply(batch_words, function(w) grepl(w, tolower(nm), fixed = TRUE))))
]

batch_cols

```

# visuals proteomics

```{r, message=FALSE, warning=FALSE, echo=TRUE}

dfp <- data.frame(
  PC1 = pcs_prot$PC1,
  PC2 = pcs_prot$PC2,
  type = clin$type_of_analyzed_samples
)

dfp$type <- as.character(dfp$type)
dfp$type[is.na(dfp$type) | dfp$type == ""] <- "NA"
dfp$type <- factor(dfp$type)

ggplot(dfp, aes(x = PC1, y = PC2, colour = type)) +
  geom_point(size = 2) +
  labs(
    title = "Proteomics PCA coloured by type of analysed_samples",
    colour = "type_of_analyzed_samples"
  ) +
  theme_bw()

```
# Visuals- transcriptomics

```{r, message=FALSE, warning=FALSE, echo=TRUE}

dfr <- data.frame(
  PC1 = pcs_rna$PC1,
  PC2 = pcs_rna$PC2,
  type = clin$type_of_analyzed_samples
)

dfr$type <- as.character(dfr$type)
dfr$type[is.na(dfr$type) | dfr$type == ""] <- "NA"
dfr$type <- factor(dfr$type)

ggplot(dfr, aes(x = PC1, y = PC2, colour = type)) +
  geom_point(size = 2) +
  labs(
    title = "RNA PCA coloured by type_of analysed samples",
    colour = "type_of_analyzed_samples"
  ) +
  theme_bw()


```


```{r, message=FALSE, warning=FALSE, echo=TRUE}

sessionInfo()


```



